src/main/java/build/buildfarm/worker/PutOperationStage.java

```diff
@@ -27,5 +36,71 @@ public PutOperationStage(InterruptingConsumer<Operation> onPut) {
   @Override
   public void put(OperationContext operationContext) throws InterruptedException {
     onPut.acceptInterruptibly(operationContext.operation);
+    synchronized (this) {
+      if (startToCount) {
+        computeAverageTimeCostPerStage(operationContext);
+        operationCount++;
+      }
+    }
+  }
+
+  public synchronized int getOperationCount() {
+    startToCount = true;
+    int currentCount = operationCount;
+    operationCount = 0;
+    return currentCount;
+  }
+
+  public synchronized float[] getAverageTimeCostPerStage() {
+    float[] currentOperationAverageTimes = averageTimeCostPerStage;
+    averageTimeCostPerStage = new float[NumOfFieldInOperationTimes];
+    return currentOperationAverageTimes;
+  }
+
+  @GuardedBy("this")
+  private void computeAverageTimeCostPerStage(OperationContext context) {
+    ExecutedActionMetadata metadata =
+        context.executeResponse.build().getResult().getExecutionMetadata();
+    Timestamp[] timestamps =
+        new Timestamp[] {
+          metadata.getQueuedTimestamp(),
+          metadata.getWorkerStartTimestamp(),
+          metadata.getInputFetchStartTimestamp(),
+          metadata.getInputFetchCompletedTimestamp(),
+          metadata.getExecutionStartTimestamp(),
+          metadata.getExecutionCompletedTimestamp(),
+          metadata.getOutputUploadStartTimestamp(),
+          metadata.getOutputUploadCompletedTimestamp(),
+        };
+
+    // The time unit we want is millisecond.
+    // 1 second = 1000 milliseconds
+    // 1 millisecond = 1000,000 nanoseconds
+    double[] times = new double[timestamps.length];
+    for (int i = 0; i < times.length; i++) {
+      times[i] =
+          (timestamps[i].getSeconds() - timestamps[0].getSeconds()) * 1000.0
+              + timestamps[i].getNanos() / (1000.0 * 1000.0);
```

Comment:
this duration computation is a bit off. You're comparing everything to the queued timestamp, but only for the seconds fields... do you want the computation to be for each sequence? The use of an array here is pretty confusing as well, does each stage correspond to an index? I think you want to do this with a structured field (with names for each logical stage).

Reply:
That's what i did on purpose. The reason is that `getSeconds()` give seconds passed since `1970-01-01`, while `getNanos()` only give `Non-negative fractions of a second at nanosecond resolution`. therefore, `getSeconds() * 1000.0 + getNanos() / 1000,000.0` will make a real timestamp. However, this will be a problem for `double`  or `float ` representation, since getSeconds() * 1000.0 is a larger number, which will make the time differences between two stages too small to be represented by either double or float number. Therefore, what i get is that operation time costs on each stage are all `0.0`s. 

To avoid that, i just use the queued time as the origin of time in this small world .